// vite-plugin-radar/dist/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  VitePluginRadar: () => VitePluginRadar,
  default: () => src_default
});
module.exports = __toCommonJS(src_exports);

// src/google-analytics.ts
var GTagSource = "https://www.googletagmanager.com";
var GTagBase = (source) => `${source}/gtag/js`;
function injectTag(options) {
  const tags = [];
  let properties = [];
  if (Array.isArray(options)) {
    properties.push(
      ...options
    );
  } else {
    properties.push(options);
  }
  properties = properties.filter((property) => Boolean(property.id));
  if (!properties.length)
    return tags;
  const mainProperty = properties.shift();
  if (!mainProperty)
    return tags;
  let template = "";
  if (mainProperty.disable)
    template += `window['ga-disable-${mainProperty.id}'] = true;
`;
  for (const property of properties) {
    if (property.disable)
      template += `window['ga-disable-${property.id}'] = true;
`;
  }
  template += "window.dataLayer = window.dataLayer || [];\n";
  template += "function gtag(){dataLayer.push(arguments);}\n";
  template += "gtag('js', new Date());\n";
  if (mainProperty.consentDefaults)
    template += `gtag('consent', 'default', ${JSON.stringify(mainProperty.consentDefaults)});
`;
  for (const property of [mainProperty, ...properties]) {
    if (property.config || property.persistentValues) {
      const config = Object.assign(
        {},
        property.persistentValues,
        property.config
      );
      template += `gtag('config', '${property.id}', ${JSON.stringify(config)});
`;
    } else {
      template += `gtag('config', '${property.id}');
`;
    }
  }
  tags.push({
    tag: "script",
    attrs: {
      src: `${GTagBase((mainProperty == null ? void 0 : mainProperty.source) || GTagSource)}?id=${mainProperty.id}`,
      async: true
    },
    injectTo: mainProperty.injectTo
  });
  tags.push({
    tag: "script",
    children: template,
    injectTo: mainProperty.injectTo
  });
  return tags;
}
var google_analytics_default = injectTag;

// src/google-tag-manager.ts
var defaultOptions = {
  gtmBase: "https://www.googletagmanager.com/gtm.js",
  nsBase: "https://www.googletagmanager.com/ns.html"
};
function injectTag2(options) {
  const tags = [];
  let properties = [];
  if (Array.isArray(options)) {
    properties.push(
      ...options.map((options2) => __spreadValues(__spreadValues({}, defaultOptions), options2))
    );
  } else {
    properties.push(__spreadValues(__spreadValues({}, defaultOptions), options));
  }
  properties = properties.filter((property) => Boolean(property.id));
  if (!properties.length)
    return tags;
  let template = "";
  template += "window.dataLayer = window.dataLayer || [];\n";
  template += "window.dataLayer.push({";
  template += "'gtm.start': new Date().getTime(),";
  template += "event:'gtm.js'";
  template += "});\n";
  tags.push({
    tag: "script",
    children: template
  });
  for (const property of properties) {
    const environmentAttachment = property.environment ? `&gtm_auth=${property.environment.auth}&gtm_preview=${property.environment.preview}` : "";
    tags.push({
      tag: "script",
      attrs: {
        src: `${property.gtmBase}?id=${property.id}${environmentAttachment}`,
        async: true
      }
    });
    tags.push({
      tag: "noscript",
      injectTo: "body-prepend",
      children: `<iframe src="${property.nsBase}?id=${property.id}${environmentAttachment}" height="0" width="0" style="display:none;visibility:hidden"></iframe>`
    });
  }
  return tags;
}
var google_tag_manager_default = injectTag2;

// src/facebook-pixel.ts
var PixelBase = "https://connect.facebook.net/en_US/fbevents.js";
var NoScriptBase = "https://www.facebook.com/tr";
function injectTag3(options) {
  const tags = [];
  let properties = [];
  if (Array.isArray(options)) {
    properties.push(
      ...options
    );
  } else {
    properties.push(options);
  }
  properties = properties.filter((property) => Boolean(property.id));
  if (!properties.length)
    return tags;
  let template = "";
  let noscriptTemplate = "";
  template += "!function(f,b,e,v,n,t,s)";
  template += "{if(f.fbq)return;n=f.fbq=function(){n.callMethod?";
  template += "n.callMethod.apply(n,arguments):n.queue.push(arguments)};";
  template += "if(!f._fbq)f._fbq=n;n.push=n;n.loaded=!0;n.version='2.0';";
  template += "n.queue=[];t=b.createElement(e);t.async=!0;";
  template += "t.src=v;s=b.getElementsByTagName(e)[0];";
  template += "s.parentNode.insertBefore(t,s)}(window, document,'script',";
  template += `'${PixelBase}');`;
  for (const property of properties) {
    template += `fbq('init', '${property.id}');`;
    noscriptTemplate += `<img height="1" width="1" style="display:none" src="${NoScriptBase}?id=${property.id}&ev=PageView&noscript=1"/>
`;
  }
  template += "fbq('track', 'PageView');";
  tags.push({
    tag: "script",
    children: template
  });
  tags.push({
    tag: "noscript",
    injectTo: "body-prepend",
    children: noscriptTemplate
  });
  return tags;
}
var facebook_pixel_default = injectTag3;

// src/vk-retargeting.ts
var RetargetingBase = "https://vk.com/js/api/openapi.js?169";
var NoScriptBase2 = "https://vk.com/rtrg";
function injectTag4(options) {
  const tags = [];
  let properties = [];
  if (Array.isArray(options)) {
    properties.push(
      ...options
    );
  } else {
    properties.push(options);
  }
  properties = properties.filter((property) => Boolean(property.id));
  if (!properties.length)
    return tags;
  let template = "";
  let noscriptTemplate = "";
  template += "!(function()";
  template += '{var t=document.createElement("script");t.type="text/javascript",';
  template += `t.async=!0,t.src="${RetargetingBase}",t.onload=function(){`;
  for (const property of properties) {
    template += `VK.Retargeting.Init("${property.id}"),`;
    noscriptTemplate += `<img height="1" width="1" style="display:none" src="${NoScriptBase2}?p=${property.id}"/>
`;
  }
  template += "VK.Retargeting.Hit()},document.head.appendChild(t);})();";
  tags.push({
    tag: "script",
    children: template
  });
  tags.push({
    tag: "noscript",
    injectTo: "body-prepend",
    children: noscriptTemplate
  });
  return tags;
}
var vk_retargeting_default = injectTag4;

// src/linkedin-insight.ts
var InsightBase = "https://snap.licdn.com/li.lms-analytics/insight.min.js";
var NoScriptBase3 = "https://px.ads.linkedin.com/collect/";
function injectTag5(options) {
  const tags = [];
  let properties = [];
  if (Array.isArray(options)) {
    properties.push(
      ...options
    );
  } else {
    properties.push(options);
  }
  properties = properties.filter((property) => Boolean(property.id));
  if (!properties.length)
    return tags;
  let template = "";
  let noscriptTemplate = "";
  template += "window._linkedin_data_partner_ids = window._linkedin_data_partner_ids || [];\n";
  for (const property of properties) {
    template += `window._linkedin_data_partner_ids.push(${property.id});
`;
    noscriptTemplate += `<img height="1" width="1" style="display:none;" alt="" src="${NoScriptBase3}?pid=${property.id}&fmt=gif" />`;
  }
  tags.push({
    tag: "script",
    children: template
  });
  tags.push({
    tag: "script",
    attrs: {
      src: `${InsightBase}`,
      async: true
    }
  });
  tags.push({
    tag: "noscript",
    injectTo: "body-prepend",
    children: noscriptTemplate
  });
  return tags;
}
var linkedin_insight_default = injectTag5;

// src/baidu-tongji.ts
var TongjiBase = "https://hm.baidu.com/hm.js";
function injectTag6(options) {
  const tags = [];
  let properties = [];
  if (Array.isArray(options)) {
    properties.push(
      ...options
    );
  } else {
    properties.push(options);
  }
  properties = properties.filter((property) => Boolean(property.id));
  if (!properties.length)
    return tags;
  let template = "";
  template += "var _hmt = _hmt || [];\n";
  template += "_hmt.push(['_setAutoPageview', true]);\n";
  for (const property of properties)
    template += `_hmt.push(['_setAccount', '${property.id}']);
`;
  tags.push({
    tag: "script",
    children: template
  });
  for (const property of properties) {
    tags.push({
      tag: "script",
      attrs: {
        src: `${TongjiBase}?${property.id}`,
        async: true
      }
    });
  }
  return tags;
}
var baidu_tongji_default = injectTag6;

// src/yandex-metrica.ts
var MetricaBase = "https://cdn.jsdelivr.net/npm/yandex-metrica-watch/tag.js";
var NoScriptBase4 = "https://mc.yandex.ru/watch/";
function injectTag7(options) {
  const tags = [];
  let properties = [];
  if (Array.isArray(options)) {
    properties.push(
      ...options
    );
  } else {
    properties.push(options);
  }
  properties = properties.filter((property) => Boolean(property.id));
  if (!properties.length)
    return tags;
  let template = "";
  let noscriptTemplate = "";
  template += "(function(m,e,t,r,i,k,a){m[i]=m[i]||function(){(m[i].a=m[i].a||[]).push(arguments)};";
  template += "m[i].l=1*new Date();k=e.createElement(t),a=e.getElementsByTagName(t)[0],k.async=1,k.src=r,a.parentNode.insertBefore(k,a)})";
  template += `(window, document, "script", "${MetricaBase}", "ym");`;
  for (const property of properties) {
    if (property.config)
      template += `ym(${property.id}, "init", ${JSON.stringify(property.config)});
`;
    else
      template += `ym(${property.id}, "init");
`;
    noscriptTemplate += `<img src="${NoScriptBase4}${property.id}" style="position:absolute;left:-9999px;" alt="" />`;
  }
  tags.push({
    tag: "script",
    children: template
  });
  tags.push({
    tag: "noscript",
    injectTo: "body-prepend",
    children: `<div>${noscriptTemplate}</div>`
  });
  return tags;
}
var yandex_metrica_default = injectTag7;

// src/microsoft-advertising.ts
var UetBase = "//bat.bing.com/bat.js";
function injectTag8(property) {
  const tags = [];
  if (!property.id)
    return tags;
  let template = "";
  template += "(function (w, d, t, r, u) {";
  template += `var f, n, i; w[u] = w[u] || [], f = function () { var o = { ti: "${property.id}" };`;
  template += 'o.q = w[u], w[u] = new UET(o), w[u].push("pageLoad") }, n = d.createElement(t), ';
  template += "n.src = r, n.async = 1, n.onload = n.onreadystatechange = function () { ";
  template += 'var s = this.readyState; s && s !== "loaded" && s !== "complete" || (f(), ';
  template += "n.onload = n.onreadystatechange = null) }, i = d.getElementsByTagName(t)[0], ";
  template += "i.parentNode.insertBefore(n, i) })";
  template += `(window, document, "script", "${UetBase}", "uetq");`;
  tags.push({
    tag: "script",
    children: template
  });
  return tags;
}
var microsoft_advertising_default = injectTag8;

// src/hotjar.ts
var HotjarBase = "https://static.hotjar.com/c/hotjar-";
function injectTag9(options) {
  const tags = [];
  if (!options.id)
    return tags;
  let template = "";
  template += "(function(h,o,t,j,a,r){";
  template += "h.hj=h.hj||function(){(h.hj.q=h.hj.q||[]).push(arguments)};";
  template += `h._hjSettings={hjid:${options.id},hjsv:6};`;
  template += 'a=o.getElementsByTagName("head")[0];';
  template += 'r=o.createElement("script");r.async=1;';
  template += "r.src=t+h._hjSettings.hjid+j+h._hjSettings.hjsv;";
  template += "a.appendChild(r);";
  template += `})(window,document,'${HotjarBase}','.js?sv=');`;
  tags.push({
    tag: "script",
    children: template
  });
  return tags;
}
var hotjar_default = injectTag9;

// src/full-story.ts
var defaultOptions2 = {
  host: "fullstory.com",
  script: "edge.fullstory.com/s/fs.js",
  namespace: "FS"
};
function injectTag10(options) {
  const tags = [];
  const fullOptions = __spreadValues(__spreadValues({}, defaultOptions2), options);
  if (!fullOptions.org)
    return tags;
  let template = "";
  template += `window['_fs_host'] = '${fullOptions.host}';`;
  template += `window['_fs_script'] = '${fullOptions.script}';`;
  template += `window['_fs_org'] = '${fullOptions.org}';`;
  template += `window['_fs_namespace'] = '${fullOptions.namespace}';`;
  template += "(function(m,n,e,t,l,o,g,y){";
  template += `if (e in m) {if(m.console && m.console.log) { m.console.log('FullStory namespace conflict. Please set window["_fs_namespace"].');} return;}`;
  template += "g=m[e]=function(a,b,s){g.q?g.q.push([a,b,s]):g._api(a,b,s);};g.q=[];";
  template += "o=n.createElement(t);o.async=1;o.crossOrigin='anonymous';o.src='https://'+_fs_script;";
  template += "y=n.getElementsByTagName(t)[0];y.parentNode.insertBefore(o,y);";
  template += "g.identify=function(i,v,s){g(l,{uid:i},s);if(v)g(l,v,s)};g.setUserVars=function(v,s){g(l,v,s)};g.event=function(i,v,s){g('event',{n:i,p:v},s)};";
  template += "g.anonymize=function(){g.identify(!!0)};";
  template += 'g.shutdown=function(){g("rec",!1)};g.restart=function(){g("rec",!0)};';
  template += 'g.log = function(a,b){g("log",[a,b])};';
  template += 'g.consent=function(a){g("consent",!arguments.length||a)};';
  template += "g.identifyAccount=function(i,v){o='account';v=v||{};v.acctId=i;g(o,v)};";
  template += "g.clearUserCookie=function(){};";
  template += "g.setVars=function(n, p){g('setVars',[n,p]);};";
  template += "g._w={};y='XMLHttpRequest';g._w[y]=m[y];y='fetch';g._w[y]=m[y];";
  template += "if(m[y])m[y]=function(){return g._w[y].apply(this,arguments)};";
  template += 'g._v="1.3.0";';
  template += "})(window,document,window['_fs_namespace'],'script','user');";
  tags.push({
    tag: "script",
    children: template
  });
  return tags;
}
var full_story_default = injectTag10;

// src/unbounce.ts
var UnbounceBase = "d3pkntwtp2ukl5.cloudfront.net/uba.js";
function injectTag11(options) {
  const tags = [];
  const sourceLocation = options === true ? UnbounceBase : options.script;
  let template = "";
  template += "var _ubaq = _ubaq || [];";
  template += '_ubaq.push(["trackGoal", "convert"]);';
  template += "(function () {";
  template += 'var ub_script = document.createElement("script");';
  template += 'ub_script.type = "text/javascript";';
  template += `ub_script.src = ("https:" == document.location.protocol ? "https://" : "http://") + "${sourceLocation}";`;
  template += 'var s = document.getElementsByTagName("script")[0];';
  template += "s.parentNode.insertBefore(ub_script, s);";
  template += "})();";
  tags.push({
    tag: "script",
    children: template
  });
  return tags;
}
var unbounce_default = injectTag11;

// src/tiktok-pixel.ts
var PixelMethods = ["page", "track", "identify", "instances", "debug", "on", "off", "once", "ready", "alias", "group", "enableCookie", "disableCookie"];
var PixelMethodsAsString = PixelMethods.map((s) => `"${s}"`).join(", ");
var TikTokPixelBase = "analytics.tiktok.com/i18n/pixel/events.js";
function injectTag12(options) {
  var _a;
  const tags = [];
  const sourceLocation = (_a = options.script) != null ? _a : TikTokPixelBase;
  if (!options.id)
    return tags;
  let template = "";
  template += "!(function (w, d, t) {";
  template += "w.TiktokAnalyticsObject = t;";
  template += "var ttq = (w[t] = w[t] || []);";
  template += `(ttq.methods = [${PixelMethodsAsString}]),`;
  template += "(ttq.setAndDefer = function (t, e) {";
  template += "t[e] = function () {";
  template += "t.push([e].concat(Array.prototype.slice.call(arguments, 0)));";
  template += "};";
  template += "});";
  template += "for (var i = 0; i < ttq.methods.length; i++) ttq.setAndDefer(ttq, ttq.methods[i]);";
  template += "(ttq.instance = function (t) {";
  template += "for (var e = ttq._i[t] || [], n = 0; n < ttq.methods.length; n++) ttq.setAndDefer(e, ttq.methods[n]);";
  template += "return e;";
  template += "}),";
  template += "(ttq.load = function (e, n) {";
  template += `var i = ("https:" == document.location.protocol ? "https://" : "http://") + "${sourceLocation}";`;
  template += "(ttq._i = ttq._i || {}), (ttq._i[e] = []), (ttq._i[e]._u = i), (ttq._t = ttq._t || {}), (ttq._t[e] = +new Date()), (ttq._o = ttq._o || {}), (ttq._o[e] = n || {});";
  template += 'var o = document.createElement("script");';
  template += '(o.type = "text/javascript"), (o.async = !0), (o.src = i + "?sdkid=" + e + "&lib=" + t);';
  template += 'var a = document.getElementsByTagName("script")[0];';
  template += "a.parentNode.insertBefore(o, a);";
  template += "});";
  template += `ttq.load("${options.id}");`;
  template += "ttq.page();";
  template += '})(window, document, "ttq");';
  tags.push({
    tag: "script",
    children: template
  });
  return tags;
}
var tiktok_pixel_default = injectTag12;

// src/simple-analytics.ts
var SimpleAnalyticsBase = "https://scripts.simpleanalyticscdn.com/latest.js";
var NoScriptBase5 = "https://queue.simpleanalyticscdn.com/noscript.gif";
function injectTag13(options) {
  const tags = [];
  if (!options.enabled)
    return tags;
  let noScriptUrl = options.noScript ? options.noScript : NoScriptBase5;
  const scriptAttrs = {
    src: options.script ? options.script : SimpleAnalyticsBase,
    async: true,
    defer: true
  };
  if (options.hostname) {
    noScriptUrl += `?hostname=${options.hostname}`;
    scriptAttrs["data-hostname"] = options.hostname;
  }
  const noscriptTemplate = `<img src="${noScriptUrl}" referrerpolicy="no-referrer-when-downgrade" alt="" />`;
  tags.push({
    tag: "script",
    injectTo: "body",
    attrs: scriptAttrs
  });
  tags.push({
    tag: "noscript",
    injectTo: "body-prepend",
    children: noscriptTemplate
  });
  return tags;
}
var simple_analytics_default = injectTag13;

// src/index.ts
function VitePluginRadar({
  enableDev = false,
  analytics,
  gtm,
  pixel,
  linkedin,
  tongji,
  metrica,
  microsoft,
  retargeting,
  hotjar,
  fullStory,
  unbounce,
  tiktok,
  simpleanalytics
}) {
  let viteConfig;
  return {
    name: "vite-plugin-Radar",
    configResolved(resolvedConfig) {
      viteConfig = resolvedConfig;
    },
    transformIndexHtml() {
      const tags = [];
      if (viteConfig.command === "serve" && !enableDev)
        return tags;
      if (analytics)
        tags.push(...google_analytics_default(analytics));
      if (gtm)
        tags.push(...google_tag_manager_default(gtm));
      if (pixel)
        tags.push(...facebook_pixel_default(pixel));
      if (retargeting)
        tags.push(...vk_retargeting_default(retargeting));
      if (tongji)
        tags.push(...baidu_tongji_default(tongji));
      if (linkedin)
        tags.push(...linkedin_insight_default(linkedin));
      if (metrica)
        tags.push(...yandex_metrica_default(metrica));
      if (microsoft)
        tags.push(...microsoft_advertising_default(microsoft));
      if (hotjar)
        tags.push(...hotjar_default(hotjar));
      if (fullStory)
        tags.push(...full_story_default(fullStory));
      if (unbounce && (unbounce === true || unbounce.enabled === true))
        tags.push(...unbounce_default(unbounce));
      if (tiktok)
        tags.push(...tiktok_pixel_default(tiktok));
      if (simpleanalytics)
        tags.push(...simple_analytics_default(simpleanalytics));
      return tags;
    }
  };
}
var src_default = VitePluginRadar;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  VitePluginRadar
});
