// wagmi/dist/esm/exports/index.js
////////////////////////////////////////////////////////////////////////////////
// Context
////////////////////////////////////////////////////////////////////////////////
export { WagmiContext, WagmiProvider, 
/** @deprecated Use `WagmiContext` instead */
WagmiContext as Context, 
/** @deprecated Use `WagmiProvider` instead */
WagmiProvider as WagmiConfig, } from '../context.js';
////////////////////////////////////////////////////////////////////////////////
// Errors
////////////////////////////////////////////////////////////////////////////////
export { BaseError } from '../errors/base.js';
export { WagmiProviderNotFoundError, } from '../errors/context.js';
////////////////////////////////////////////////////////////////////////////////
// Hooks
////////////////////////////////////////////////////////////////////////////////
export { useAccount, } from '../hooks/useAccount.js';
export { useAccountEffect, } from '../hooks/useAccountEffect.js';
export { useBalance, } from '../hooks/useBalance.js';
export { useBlock, } from '../hooks/useBlock.js';
export { useBlockNumber, } from '../hooks/useBlockNumber.js';
export { useBlockTransactionCount, } from '../hooks/useBlockTransactionCount.js';
export { useBytecode, } from '../hooks/useBytecode.js';
export { useCall, } from '../hooks/useCall.js';
export { useChainId, } from '../hooks/useChainId.js';
export { useChains, } from '../hooks/useChains.js';
export { useClient, } from '../hooks/useClient.js';
export { useConfig, } from '../hooks/useConfig.js';
export { useConnect, } from '../hooks/useConnect.js';
export { useConnections, } from '../hooks/useConnections.js';
export { useConnectors, } from '../hooks/useConnectors.js';
export { useConnectorClient, } from '../hooks/useConnectorClient.js';
export { useDisconnect, } from '../hooks/useDisconnect.js';
export { useEnsAddress, } from '../hooks/useEnsAddress.js';
export { useEnsAvatar, } from '../hooks/useEnsAvatar.js';
export { useEnsName, } from '../hooks/useEnsName.js';
export { useEnsResolver, } from '../hooks/useEnsResolver.js';
export { useEnsText, } from '../hooks/useEnsText.js';
export { useEstimateFeesPerGas, 
/** @deprecated Use `useEstimateFeesPerGas` instead */
useEstimateFeesPerGas as useFeeData, } from '../hooks/useEstimateFeesPerGas.js';
export { useEstimateGas, } from '../hooks/useEstimateGas.js';
export { useEstimateMaxPriorityFeePerGas, } from '../hooks/useEstimateMaxPriorityFeePerGas.js';
export { useFeeHistory, } from '../hooks/useFeeHistory.js';
export { useGasPrice, } from '../hooks/useGasPrice.js';
export { useInfiniteReadContracts, 
/** @deprecated Use `useInfiniteReadContracts` instead */
useInfiniteReadContracts as useContractInfiniteReads, } from '../hooks/useInfiniteReadContracts.js';
export { usePrepareTransactionRequest, } from '../hooks/usePrepareTransactionRequest.js';
export { useProof, } from '../hooks/useProof.js';
export { usePublicClient, } from '../hooks/usePublicClient.js';
export { useReadContract, 
/** @deprecated Use `useWriteContract` instead */
useReadContract as useContractRead, } from '../hooks/useReadContract.js';
export { useReadContracts, 
/** @deprecated Use `useWriteContract` instead */
useReadContracts as useContractReads, } from '../hooks/useReadContracts.js';
export { useReconnect, } from '../hooks/useReconnect.js';
export { useSendTransaction, } from '../hooks/useSendTransaction.js';
export { useSignMessage, } from '../hooks/useSignMessage.js';
export { useSignTypedData, } from '../hooks/useSignTypedData.js';
export { useSimulateContract, } from '../hooks/useSimulateContract.js';
export { useStorageAt, } from '../hooks/useStorageAt.js';
export { useSwitchAccount, } from '../hooks/useSwitchAccount.js';
export { useSwitchChain, } from '../hooks/useSwitchChain.js';
export { 
/** @deprecated Use `useReadContracts` instead */
useToken, } from '../hooks/useToken.js';
export { useTransaction, } from '../hooks/useTransaction.js';
export { useTransactionConfirmations, } from '../hooks/useTransactionConfirmations.js';
export { useTransactionCount, } from '../hooks/useTransactionCount.js';
export { useTransactionReceipt, } from '../hooks/useTransactionReceipt.js';
export { useVerifyMessage, } from '../hooks/useVerifyMessage.js';
export { useVerifyTypedData, } from '../hooks/useVerifyTypedData.js';
export { useWalletClient, } from '../hooks/useWalletClient.js';
export { useWaitForTransactionReceipt, } from '../hooks/useWaitForTransactionReceipt.js';
export { useWatchBlocks, } from '../hooks/useWatchBlocks.js';
export { useWatchBlockNumber, } from '../hooks/useWatchBlockNumber.js';
export { useWatchContractEvent, } from '../hooks/useWatchContractEvent.js';
export { useWatchPendingTransactions, } from '../hooks/useWatchPendingTransactions.js';
export { useWriteContract, 
/** @deprecated Use `useWriteContract` instead */
useWriteContract as useContractWrite, } from '../hooks/useWriteContract.js';
////////////////////////////////////////////////////////////////////////////////
// Hydrate
////////////////////////////////////////////////////////////////////////////////
export { Hydrate, } from '../hydrate.js';
////////////////////////////////////////////////////////////////////////////////
// @wagmi/core
////////////////////////////////////////////////////////////////////////////////
export { createConfig, createConnector, ChainNotConfiguredError, ConnectorAlreadyConnectedError, ConnectorNotFoundError, ConnectorAccountNotFoundError, ProviderNotFoundError, SwitchChainNotSupportedError, createStorage, noopStorage, 
// Transports
custom, fallback, http, webSocket, unstable_connector, 
// Utilities
cookieStorage, cookieToInitialState, deepEqual, deserialize, normalizeChainId, parseCookie, serialize, } from '@wagmi/core';
////////////////////////////////////////////////////////////////////////////////
// Version
////////////////////////////////////////////////////////////////////////////////
export { version } from '../version.js';
