// @graphql-mesh/http/esm/graphqlHandler.js
import { createYoga, useLogger } from 'graphql-yoga';
export const graphqlHandler = ({ getBuiltMesh, playgroundTitle, playgroundEnabled, graphqlEndpoint, corsConfig, batchingLimit, }) => {
    let yoga;
    let yoga$;
    return (request, ctx) => {
        if (yoga) {
            return yoga.handleRequest(request, ctx);
        }
        if (!yoga$) {
            yoga$ = getBuiltMesh().then(mesh => {
                yoga = createYoga({
                    plugins: [
                        ...mesh.plugins,
                        useLogger({
                            skipIntrospection: true,
                            logFn: (eventName, { args }) => {
                                if (eventName.endsWith('-start')) {
                                    mesh.logger.debug(`\t headers: `, args.contextValue.headers);
                                }
                            },
                        }),
                    ],
                    logging: mesh.logger,
                    maskedErrors: false,
                    graphiql: playgroundEnabled && {
                        title: playgroundTitle,
                    },
                    cors: corsConfig,
                    graphqlEndpoint,
                    landingPage: false,
                    batching: batchingLimit ? { limit: batchingLimit } : false,
                });
                return yoga;
            });
        }
        return yoga$.then(yoga => yoga.handleRequest(request, ctx));
    };
};
